src/iterators.rs

```diff
@@ -471,14 +478,32 @@ macro_rules! custom_vec_iter_impl {
                 Ok(self.$data.len())
             }
 
-            fn __getitem__(&self, idx: isize) -> PyResult<$T> {
-                if idx.abs() >= self.$data.len().try_into().unwrap() {
-                    Err(PyIndexError::new_err(format!("Invalid index, {}", idx)))
-                } else if idx < 0 {
-                    let len = self.$data.len();
-                    Ok(self.$data[len - idx.abs() as usize].clone())
-                } else {
-                    Ok(self.$data[idx as usize].clone())
+            fn __getitem__(&self, py: Python, idx: SliceOrInt) -> PyResult<PyObject> {
+                match idx {
+                    SliceOrInt::Slice(slc) => {
+                        let len = self.$data.len().try_into().unwrap();
+                        let indices = slc.indices(len)?;
+                        let start: usize = indices.start.try_into().unwrap();
+                        let stop: usize = indices.stop.try_into().unwrap();
+                        let step: usize = indices.step.try_into().unwrap();
+                        let return_vec = $name {
+                            $data: (start..stop)
+                                .step_by(step)
+                                .map(|i| self.$data[i].clone())
+                                .collect(),
+                        };
+                        Ok(return_vec.into_py(py))
+                    }
+                    SliceOrInt::Int(idx) => {
+                        if idx.abs() >= self.$data.len().try_into().unwrap() {
```

Comment:
It's my fault but this check should have been:

```suggestion
                        let len = self.$data.len() as isize;
                        if idx >= len || idx < -len {
```

since it's valid to access the first element of a sequence `xs` of length `len` as `xs[-len]` (previously we'd wrongly raise an index error).